openfl/transport/grpc/director_client.py

```diff
@@ -0,0 +1,280 @@
+        """Send envoy health check."""
+        status = director_pb2.CollaboratorStatus(
+            name=collaborator_name,
+            is_experiment_running=is_experiment_running,
+        )
+        status.valid_duration.seconds = valid_duration
+        logger.debug(f'Sending health check status: {status}')
+
+        return self.stub.CollaboratorHealthCheck(status)
+
+
+class DirectorClient:
+    """Director client class for users."""
+
+    def __init__(self, client_id, director_uri, disable_tls, root_ca, key, cert) -> None:
+        """Initialize director client object."""
+        channel_opt = [('grpc.max_send_message_length', 512 * 1024 * 1024),
+                       ('grpc.max_receive_message_length', 512 * 1024 * 1024)]
+        if disable_tls:
+            channel = grpc.insecure_channel(director_uri, options=channel_opt)
+        else:
+            with open(root_ca, 'rb') as f:
+                root_ca_b = f.read()
+            with open(key, 'rb') as f:
+                key_b = f.read()
+            with open(cert, 'rb') as f:
+                cert_b = f.read()
+
+            credentials = grpc.ssl_channel_credentials(
+                root_certificates=root_ca_b,
+                private_key=key_b,
+                certificate_chain=cert_b
+            )
+
+            channel = grpc.secure_channel(director_uri, credentials, options=channel_opt)
+        self.stub = director_pb2_grpc.FederationDirectorStub(channel)
+
+        self.client_id = client_id
+        self.header = director_pb2.RequestHeader(sender=self.client_id)
+
+    def report_shard_info(self, shard_descriptor) -> bool:
+        """Report shard info to the director."""
+        logger.info('Send report AcknowledgeShard')
+
+    def set_new_experiment(self, name, col_names, arch_path,
+                           initial_tensor_dict=None):
+        """Send the new experiment to director to launch."""
+        logger.info('SetNewExperiment')
+        model_proto = None
+        if initial_tensor_dict:
+            model_proto = construct_model_proto(initial_tensor_dict, 0, NoCompressionPipeline())
+
+        with open(arch_path, 'rb') as arch:
+            def st():
+                max_buffer_size = (2 * 1024 * 1024)
+                chunk = arch.read(max_buffer_size)
+                while chunk != b'':
+                    if not chunk:
+                        raise StopIteration
+                    # TODO: add hash or/and size to check
+                    experiment_info = director_pb2.ExperimentInfo(
+                        header=self.header,
+                        name=name,
+                        collaborator_names=col_names,
+                        model_proto=model_proto
+                    )
+                    experiment_info.experiment_data.size = len(chunk)
+                    experiment_info.experiment_data.npbytes = chunk
+                    yield experiment_info
+                    chunk = arch.read(max_buffer_size)
```

Comment:
Maybe better to create stand alone function(generator) which will read the archive than use closure here.

Reply:
Could you please create issue for refactoring.